Add support for selecting schemas based on contextual "tags", with various mechanisms to set those tags

radix opened the pull request:
## Changes

This PR introduces the concept of "tags" on the Context object, which are key/value settings.

It also provides various new schema directives which make use of those tags.

1. `"set_tag": {"tag_name": "foo", "value": "bar"}` - this sets the tag `foo` to the fixed value `bar`.
2. `"set_tag": {"tag_name": "foo", "key": "abc"}` - this sets the tag `foo` to the value `obj["abc"]` -- so, it assumes that the current object being validated is a dict.
3. `"set_tag": "foo"` - this sets the tag `foo` to the value of `obj["foo"]` -- it's just a shorthand for the first form where the tag name and the dict key are identical.
4. `"hook_context": lambda value, context: context.set_tag("mytag", value[0])` - this sets the tag `mytag` to the value `value[0]`, so it assumes the value being validated is a list.

`hook_context` is the most general mechanism for setting tags -- `set_tag` is just a convenient, data-driven syntax for setting tags, but if you need to do something more fancy you can always use `hook_context`.

Those are all the directives which *set* tags. But what good are tags? This PR also introduces directives that *read* them to choose schemas.

4. `when_tag_is` -- this is identical to `when_key_is`, but it uses tags instead of looking up a key on the current object.
5. `"choose_schema": lambda value, context: schema1 if context.get_tag("foo") == "bar" else schema2`

Again, `choose_schema` is the more general mechanism for choosing a schema based on a tag -- it can do everything that `when_tag_is` can do, but `when_tag_is` can be expressed in simple data structures (which can potentially be serialized).


## What this is all about

The combination of the new directives (either `set_tag` or `hook_context`, combined with either `when_tag_is` or `choose_schema`) gives us the ability to *select schemas at a distance*.

In other words, while `when_key_is` allows you to take a value like `{"type": "foo", "foo_specific_key": "bar"}` and validate it in a way that selects a schema based on `foo`, the new mechanism allows you to do this in a deeply nested structure.

For example:

```json
{
    "type": "foo",
    "common": {...},
    "data_service": {
        "renderers": [
            {"foo_specific": "bar"}
        ]
    }
}
```

Let's assume that the `{"flavor": "bar"}` dict has a schema specific to `foo`. With these new directives, we can have a schema like the following:

```yaml
type: dict
set_tag: "type"
schema:
  type: foo
  common: {type: dict}
  data_service:
    type: dict
    schema:
      renderers:
        type: list
        schema:
          type: dict
          when_tag_is:
            tag: type
            choices:
              foo: {type: dict, schema: {foo_specific: {type: string}}
              bar: {type: dict, schema: {bar_specific: {type: integer}}
```

Notice in particular the `when_tag_is` directive deep inside the schema, while it is *referencing* a tag set by the very top of the schema -- so the `type` you choose at the top of the object is varying what schemas are used to validate the elements of the `renderers` list.

radix commented:
This doesn't have anything in the README because I am working on a full guide that I'll do in a separate PR.

chrisgemignani commented on sureberus/__init__.py:
```diff
@@ -32,6 +33,16 @@ def register_schemas(self, registry):
+        tags[tag] = value
+        return attr.evolve(self, tags=tags)
+
+    def get_tag(self, tag):
```
It seems useful to be able to supply a default if tags has not been set.

```
def get_tag(self, tag, default=TAG_DEFAULT_NOT_SET):
    if default is TAG_DEFAULT_NOT_SET:
        if tag not in self.tags:
            raise E.TagNotFound(tag, self.tags.keys(), self.stack)
        return self.tags[tag]
    else:
        return self.tags.get(tag, default)
```

radix replied:
Gotcha. I've also added a `default_choice` to `when_tag_is` to match the `default_choice` in `when_key_is`, to take advantage of this.

radix commented:
Thanks for the comments @chrisgemignani!

I've

1. renamed `hook_context` to `modify_context`
2. added a `default_choice` to `when_tag_is`, and added the default argument to `get_tag`.

I am working on some ideas for an improvement to the schema language that would unify all the choose-schema stuff (`when_key_is`, `when_key_exists`, `when_tag_is`, and `choose_schema`). I'll tinker around with it for a bit, but I might just end up adding this directive as-is if I can't figure that out quickly enough.

radix commented:
@chrisgemignani I've committed a new syntax for `when_tag_is`. It's now:

```json
{
"choose_schema": {
    "when_tag_is": {
        "tag": "foo",
        "choices": {...}
    }
}
}
```

or, using the Python syntax,

```python
S.Dict(
    choose_schema=S.when_tag_is("foo", {...})
)
```

This makes `choose_schema` the central directive which allows selecting a schema to apply.

I'd like to expose `when_key_is` and `when_key_exists` as sub-directives of `choose_schema` as well, but it wasn't trivial so I'll do that in another PR (and maintain backwards compatibility, of course).

stevenwally commented:
This looks good to me!

That said, I would also love to have a guided tour / conversation with you sometime, because I am not sure I understand what the higher level implications are that come with the introduction of tags and what use-cases would look like. I don't fully understand what _action-from-a-distance_ means.

radix commented:
## The History of `when_key_is`

Okay, let's talk about the use-case! I think it's fine to describe it here since it will probably all end up being documented at docs.juiceboxdata.com...

First, there's the general concept of variable schemas. When you're validating a document, such as a Juicebox story configuration, you sometimes have situations where the values of certain fields in the document affect which *other* fields should be available. 

For example, in our DataSource schema, we want to allow people to provide data of these two forms: (these are example data, not schemas):

```json
{"type": "sql", "connection_string": "sql://whatever", "schema": "myschema", "table": "mytable"}
```

**or**

```json
{"type": "legacy", "connection": "juicebox", "schema": "myschema", "table": "mytable"}
```

**or**

```json
{"type": "gsheets", "spreadsheet_url": "https://google.com/spreadsheet/whatever", "token": "..."}
```

(this last one is still TBD)

So, we have this dictionary that can take different forms, and we want to make sure that when you say that the type is "gsheets", that you include a "spreadsheet_url" but not a "connection_string".

Cerberus provided a mechanism for this called `anyof`. It let you specify a *list* of schemas to try. It would go through each one and validate successfully if any of them individually would apply. There were various problems with this (including that Cerberus separates normalization from validation, so the whole `anyof` mechanism was just plain crippled, but that's not super relevant for now). Maybe one of the most important problems with that is that it gave really bad error messages. With that formulation of the schema, all the validator can say is "sorry, I tried each of these schemas individually and none of them worked". Which led to very VERBOSE error messages that did a very poor job of showing you exactly what went wrong.

Ok, so none of that is really directly relevant to the current PR: it's all just history, and, led to the creation of PR https://github.com/radix/sureberus/pull/2, which is where I originally defined `when_key_is` and `when_key_exists`. These mechanisms gave a nice way to declare: "hey Sureberus, when *type* is *gsheets*, use *this* schema for the rest of the keys. When *type* is *sql*, use *this* schema". That gave us much better error messages, since Sureberus can say: oh, user, I can see that you said that the *type* is *sql*, so I can tell you exactly how your data does not match the schema associated with *sql*!

So that's all good and has been proven out, I think (barring bikeshedding on the syntax of the schemas). I think it's also just a much more principled, or at least aesthetic, way to declare the schemas. We're encoding more rich information into the schema which is allowing Sureberus to do smarter things.

## But now what's up w/ this PR??? Introducing `when_tag_is`

Ok, so we have this cool mechanism for "dynamically selecting" a schema based on data in the value, called `when_key_is`. It is currently limited to selecting a schema specifically based on the value of a dictionary-key in the value being validated. And it selects it *for that specific object*. So if you're using `when_key_is` on a `"type"` key, it means you can only choose schemas to use to validate that very dictionary which contains the "type" key. But this isn't the only situation in which we want to choose schemas dynamically.

For example! What if you had a slice configuration which was non-variable, but it contained a nested dictionary, deep down in the data structure, which varies based on what `slice_type` you selected way up at the top of the slice object?

**EXAMPLE 1**: option-chooser

```yaml
  - slice_type: option-chooser
    slug: perf_filters_emphasized1
    title: "Choose which you would like to focus on:"
    style: ...
    config: ...
    data_service:
      type: renderers
      data_source: generic
      renderers:
        - flavor: buttons
          name: button_name
          buttons:
            - date: Time Range
            - category: Category
            - name: People
            - department: Department
          group_by: "group_name"
```

**EXAMPLE 2:** 

```yaml
  - slice_type: trend
    slug: perf_trend
    title: See how the trend for <%= metric_chooser.selectionDisplay() %> has changed over time.
    style: ...
    config: ...
    data_service:
      type: renderers
      data_source: generic
      renderers:
        - flavor: ranged
          date_column: date
          data:
            metrics: ["@metric"]
            dimensions: ["date"]
            order_by: ["date"]
```

The `slice_type` field way at the top of this object doesn't really need to affect most of the schema. We just want it to affect how we parse the elements inside the `renderers` list, which is way deep down in this structure. Until now, using the `when_key_is` directive, we had to copy the whole slice schema for each slice type, with only a little bit of it changed, in each sub-schema in the `choices` field of the `when_key_is` directive. Not only was that kind of difficult code to write, it also is SUPER inefficient based on the number of slice types. I didn't actually count the size of the schema, but I wouldn't be surprised if the fully realized schema passed to sureberus was, like, a megabyte, with quite a lot of duplicated schema directives.

So I'm introducing this new directive, which allows us to put some distance between:

- what part of the data is affecting the schema (`slice_type` in the root), and
- where the schema actually varies (the elements in the `renderers` list).

That's what I mean by "select schemas at a distance".


Ok, I'm getting really tired so I'll leave it at that. Maybe reading the examples in the PR description at this point again will help drive the point home. If this isn't crystal clear yet, @meghna or @stevenwally, please let me know! I'm working on a real documentation guide for Sureberus and I'm sure that the effort I'm putting into explaining this will be very useful in producing material for it!
